### PR TITLE
[Bulk Actions] Update Bulk Actions UI

### PR DESCRIPTION
**Pull request by mrcthms**
### WHY are these changes introduced?

Related to https://github.com/Shopify/web/issues/102861

We want to improve the UX of the bulk actions and select actions in IndexTables and ResourceLists. The BulkActions should display in the sticky table header, the Pagination should now be sticky, and the SelectAllActions should behave in the same way as the pagination, sticky to the bottom of the table.

Spin URL for ResourceList: https://admin.web.bulk-ui-3.marc-thomas.eu.spin.dev/store/shop1/products/13
Spin URL for IndexTable: https://admin.web.bulk-ui-3.marc-thomas.eu.spin.dev/store/shop1/products?selectedView=all
Spin URL for ResourceListWithHeader: https://admin.web.bulk-ui-3.marc-thomas.eu.spin.dev/store/shop1/draft_orders?selectedView=all

Resume Spin link: https://spin-control.shopify.io/instances/bulk-ui-3.marc-thomas.eu.spin.dev/resume

https://github.com/Shopify/polaris/assets/2562596/372d275d-1138-451b-8457-5b3664818994

https://github.com/Shopify/polaris/assets/2562596/a80aaf5e-8f44-4ca0-9227-8debdf7d8445

### WHAT is this pull request doing?

- Transferring a lot of the logic that was in BulkActions to SelectAllActions, due to BulkActions no longer being sticky and SelectAllActions now being sticky
- Updating button styles within BulkActions to match the more condensed mode of the sticky bar
- Match SelectAllActions styling to match pagination
- Pagination is now sticky

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#local-development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

### 🎩 checklist

- [x] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [x] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [x] Updated the component's `README.md` with documentation changes
- [x] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

### CODE REVIEW

**Comment by chloerice** [comment]
Hey @mrcthms! Will give this a review tomorrow 👀

File: polaris-react/src/components/BulkActions/BulkActions.tsx [hunk 1]
```diff
@@ -41,10 +55,10 @@ export interface BulkActionsProps {
   onSelectModeToggle?(selectMode: boolean): void;
   /** Callback when more actions button is toggled */
   onMoreActionPopoverToggle?(isOpen: boolean): void;
-  /** If the BulkActions is currently sticky in view */
-  isSticky?: boolean;
-  /** The width of the BulkActions */
-  width: number;
```
Comment (sam-b-rose) [inline polaris-react/src/components/BulkActions/BulkActions.tsx:47]: Just a note that we will need to update [this bit of code in web](https://github.com/Shopify/web/blob/adeb8a4652a7f5315977d5ef58d9191b929ff3b0/app/shared/components/ResourceListWithHeader/ResourceListWithHeader.tsx#L706-L714)

**Review by sam-b-rose — COMMENTED** [review]
Looks excellent, much nicer UX imo and good refactor of the Bulk Actions!

I did notice that the bottom sticky pagination sometimes will not render in the proper location. This was happening for me on both the ResourceList and IndexTable. Seems to work sometimes but not on every reload.

https://github.com/Shopify/polaris/assets/11774595/adad6d31-cfb9-4c25-9a98-7da52cf2dbfd


https://github.com/Shopify/polaris/assets/11774595/74b0c442-1203-4af8-b72a-4efa8464d1dd

**Comment by mrcthms** [comment]
> Looks excellent, much nicer UX imo and good refactor of the Bulk Actions!
> 
> I did notice that the bottom sticky pagination sometimes will not render in the proper location. This was happening for me on both the ResourceList and IndexTable. Seems to work sometimes but not on every reload.
> 
>  index-table.mov 
>  resource-list.mov

@sam-b-rose Good spot – updated the code to fix this, thanks!

**Comment by sam-b-rose** [comment]
@mrcthms Nice, the position issue looks to be resolved 👍 

I am having trouble loading the [With Bulk Action and Selection Across Pages](https://5d559397bae39100201eedc1-ksjquwbjgk.chromatic.com/?path=/story/all-components-indextable--with-bulk-actions-and-selection-across-pages) storybook example. Seems to be a `key` error in the console, but I'm not sure that is the root issue.

File: polaris-react/src/components/BulkActions/BulkActions.tsx (hunk 1 above)
Comment (laurkim) [reply polaris-react/src/components/BulkActions/BulkActions.tsx:47]: @mrcthms is there more context for why these are being removed in a minor release and not being deprecated first and then removed in the next major release?

File: polaris-react/src/components/SelectAllActions/SelectAllActions.tsx [hunk 2]
```diff
@@ -62,32 +58,39 @@ export const SelectAllActions = forwardRef(function SelectAllActions(
...
-            {paginatedSelectAllMarkup}
+          <div className={wrapperClasses} ref={nodeRef}>
+            <Box
+              background="bg-surface-secondary"
+              borderBlockStartWidth="025"
+              borderColor="border"
+              width="100%"
+              paddingBlockStart="300"
+              paddingBlockEnd="300"
+              paddingInlineStart="400"
```
Comment (laurkim) [inline polaris-react/src/components/SelectAllActions/SelectAllActions.tsx]: It looks like the inline start padding on the checkbox is 12px, whereas the select all text inline start padding is 16px. There's a bit of inconsistency with spacing due to it but also not 100% sure if this intentional

04-49-pfx81-i584n

File: polaris-react/src/components/BulkActions/tests/BulkActions.test.tsx [hunk 3]
```diff
@@ -15,9 +16,15 @@ interface Props {
   disabled: boolean;
   width: number;
   isSticky: boolean;
```
Comment (laurkim) [inline polaris-react/src/components/BulkActions/tests/BulkActions.test.tsx]: If these props are being removed, we can remove references to them throughout this file.

File: polaris-react/src/components/IndexTable/IndexTable.stories.tsx [hunk 4]
```diff
@@ -4398,6 +4367,272 @@ export function WithPagination() {
   );
 }
 
+export function WithStickyScrollBar() {
```
Comment (laurkim) [inline polaris-react/src/components/IndexTable/IndexTable.stories.tsx:4676]: Can we update the IndexTable examples in the `polaris.shopify.com` directory to align with any updates to these storybook examples?

File: polaris-react/src/components/ResourceList/ResourceList.stories.tsx [hunk 5]
```diff
@@ -290,17 +293,26 @@ export function WithBulkActionsAndManyItems() {
...
-        resourceName={resourceName}
-        items={items}
-        renderItem={renderItem}
-        selectedItems={selectedItems}
-        onSelectionChange={setSelectedItems}
-        promotedBulkActions={promotedBulkActions}
-        bulkActions={bulkActions}
-      />
-    </Card>
+    <BlockStack gap="400">
```
Comment (laurkim) [inline polaris-react/src/components/ResourceList/ResourceList.stories.tsx:296]: Same re: updating the ResourceList examples in the `polaris.shopify.com` directory to align with any updates to these storybook examples

File: polaris-react/src/components/BulkActions/BulkActions.tsx [hunk 6]
```diff
@@ -285,9 +325,12 @@ class BulkActionsInner extends PureComponent<CombinedProps, State> {
 
     const groupContent =
       promotedActionsMarkup || actionsPopover ? (
-        <InlineStack gap="300">
-          {promotedActionsMarkup}
-          {actionsPopover}
+        <InlineStack gap="300" blockAlign="center">
+          <CheckableButton {...checkableButtonProps} />
+          <InlineStack gap="100" blockAlign="center">
```
Comment (laurkim) [inline polaris-react/src/components/BulkActions/BulkActions.tsx:330]:
```suggestion
          <InlineStack gap="200" blockAlign="center">
```

Nit: we tend to have 8px of spacing when buttons are rendered next to each other (i.e., [ButtonGroup](https://storybook.polaris.shopify.com/?path=/story/all-components-buttongroup--default)) so light suggestion to have this be consistent by updating the gap.

File: polaris-react/src/components/ResourceList/ResourceList.stories.tsx [hunk 7]
```diff
@@ -290,17 +293,26 @@ export function WithBulkActionsAndManyItems() {
...
+      <Card padding="0" roundedAbove="sm">
```
Comment (laurkim) [inline polaris-react/src/components/ResourceList/ResourceList.stories.tsx]:
```suggestion
      <Card padding="0">
```

Card was recently updated to set `roundedAbove` to `sm` by default.

**Review by laurkim — COMMENTED** [review]
@mrcthms nice refactor! 💯 

Added some small suggestions but also probably worth reviewing the new storybook baseline changes. A majority of the differences look expected due to mock data content but the IndexFilter story diffs look like they could be a valid change to look into. It looks the same when comparing your branch locally to prod but on the Chromatic diff, the table appears to have a heavier shadow.

File: polaris-react/src/components/BulkActions/BulkActions.tsx (hunk 6 above)
Comment (mrcthms) [reply polaris-react/src/components/BulkActions/BulkActions.tsx:330]: @laurkim I wouldn't want to buck the trend normally, but we have a `100` gap between the buttons and the edge of the bulk actions so having the gap between buttons as `100` helps keep balance. I've attached a screenshot of how it looks at `200` and it just feels a teeny bit off
Screenshot 2024-01-05 at 13 24 22

Comment (laurkim) [reply polaris-react/src/components/BulkActions/BulkActions.tsx:330]: @mrcthms makes sense to me, sounds good to just leave it at 100 then! 🤝

**Review by laurkim — COMMENTED** [review]
@mrcthms thanks for pushing updates!

Some small things I noticed in latest chromatic when tophatting:
- [IndexTable - With Pagination story](https://5d559397bae39100201eedc1-njwmtvynbn.chromatic.com/?path=/story/all-components-indextable--with-pagination) renders checkboxes and table headers for xs breakpoint (but doesn't in [prod](https://storybook.polaris.shopify.com/?path=/story/all-components-indextable--with-pagination)) 
- [ResourceList - With Selection And No Bulk Actions](https://5d559397bae39100201eedc1-njwmtvynbn.chromatic.com/?path=/story/all-components-resourcelist--with-selection-and-no-bulk-actions) checkbox disappears when you select any or all rows

**Review by sam-b-rose — APPROVED** [review]
Awesome work, this UX is so nice!! 🤩

**Pull request merged**